File: tests/journal.lib/error_example_nonfatal.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2023 all rights reserved


// get the journal
#include <pyre/journal.h>
// support
#include <cassert>


// type aliases
using trash_t = pyre::journal::trash_t;
using myerror_t = pyre::journal::error_t;


// basic error example with a non-fatal channel
int main() {
    // make an error channel
    myerror_t channel("tests.journal.error");

    // make it non-fatal
    channel.fatal(false);
    // send the output to the trash
    channel.device<trash_t>();

    // errors are fatal by default, so attempt to
    try {
        // inject something into the channel; no exception should be raised
        channel
            << pyre::journal::at(__HERE__)
            << pyre::journal::note("time", "now")
            << "error channel:" << pyre::journal::newline
            << "    hello world!" << pyre::journal::endl;
    // if the error triggered an exception
    } catch (const myerror_t::exception_type &) {
        // unreachable
        throw std::logic_error("unreachable");
    }

    // all done
    return 0;
}


// end of file

File: tests/journal.ext/error_example_nonfatal.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2023 all rights reserved


def test():
    """
    Exercise a non-fatal error channel with a realistic example
    """
    # get the trash can
    from journal.ext.journal import Trash as trash
    # and the channel
    from journal.ext.journal import Error as error

    # make an error channel
    channel = error(name="tests.journal.error")
    # make it non-fatal
    channel.fatal = False
    # send the output to trash
    channel.device = trash()

    # add some metadata
    channel.notes["time"] = "now"

    # carefully
    try:
        # inject
        channel.line("error channel:")
        channel.log("    hello world!")
    # if the correct exception was raised
    except channel.ApplicationError as error:
        # shouldn't get here
        assert False, "unreachable"

    # all done
    return


# main
if __name__ == "__main__":
    # prohibit the journal bindings
    journal_no_libjournal = True
    # run the test
    test()


# end of file
